### PR TITLE
Score machine

### LINKED ISSUE #5817: End-of-round-score screen B: 100
An often talked about subject is the lack of motivation for the station personnel to do anything. Antagonists have greentext, but the majority of players in a given sessions aren't explicitly rewarded or punished for good or bad behaviour. I think it would have a measureable positive effect on the game if there were objectives for the crew to accomplish.

the Station should get greentext just like antagonists, having a randomly chosen objective to accomplish. For now, it will just be filling the plasma sheet quota that is printed out at the communications console at the start of every round (track how much plasma ore has been processed by the ore smelter). Perhaps we can have objectives for other ore types as well?

### What is Expected:

- [x] Station greentext  - **added in #6878 **
- [x] Certain activities and events should be tracked and added together to generate a performance rating for the crew at the end of round. see below for a general idea of what it should look like. For a more complete understanding, refer to /tg/. Not everything that /tg/ tracks can be added at this time, but let's do what we can. There may be a few things we want to track that /tg/ doesn't as well. A discussion should be hosted on the topic in the discord.

Screenshot_2020-11-26_093425

### PR DESCRIPTION
**Pull request by MaxIsJoe**
close #5817

## The score machine is alive:

This PR adds in the score machine, a flexible system to keep track of scores of anything in-game. With this we can kick-start the development of #8933 much quicker now as one of the critical foundations of the thief role has been added.
To showcase this new system, I've created some scores to keep track of (some from SS13, others are new and original) to demonstrate how they could work and also finally add End-Of-Round scores, a feature that already exists in SS13 but has been lacking from here for a very long time.

The UI is not the main show here so corners have been cut to make the development of this faster (and to keep my sanity safe from Unity UI) but we can build off this in the future whenever we want.

## Explaining how the system works for reviewers.

The code is heavily documented and easy to read and understand, so this section is almost not needed; but just in case I'll still explain the surface of how this works.

The score machine keeps track of items by creating entries, each entry has an ID and the score object that is associated with it.
To create a new entry, simply use the function `ScoreMachine.AddNewScoreEntry()` and follow the requirements which are also documented using xml 
image

ID is the string that you will use to modify scores overtime.
scoreName is the name of the entry when it's displayed in UI
type is the type of score (more on that below)
Category defines what use-case this score can go to, `MiscScore` is the most general type of score and `AntagScore` is a thing added for anyone expanding upon scores in the future or moving antag objectives to the score system. `StationScore` are meant for giving a summary and rating of how bad or good the round went for the station and it's crew and is always displayed at the round end score screen.
alignment lets you decide if this score is considered a good thing or a bad thing.

After that you can simply modify your scores by using the `AddToScoreBool` , `AddToScoreString` and `AddToScoreInt` or your own custom function for a custom type.

what are types?
each score entry has it's own type which is identified as `ScoreEntry[Type]`, as of right now there are only types for strings, ints and bools, but you can create your own custom ones and hook them into the score machine system by updating the necessary enums and functions.
This allows us to create more complex score entry types that can hold more data and even functionality whenever we like.

At the end of the round, RoundEndScoreBuilder takes care of displaying the station's score for us and it also generates common scores entries to be tracked of at the start of the round.

## Can you add X or Y?

I'd like to not make this PR any bigger and would like for it to be reviewed quickly and merged ASAP, so I can resume work on other stuff. I'm not going to add or change anything further in this PR except for code quality changes or changes that are necessary for the bounty to be made.

However, after this is merged I'd happily add more things to keep track of and expand on this system a lot more but as it is right now I'm happy with how it is.

Branch hopping is a pain man..

# Chanelog:

CL: [New] - Added the End-Of-Round score screen.

### CODE REVIEW

**Comment by MaxIsJoe** [comment]
Question, do we **really** need a string score entry? I don't see what's the point of it.

**Comment by MaxIsJoe** [comment]
image

made a clipboard texture for the UI and got the basic stuff setup

**Comment by MaxIsJoe** [comment]
> Question, do we **really** need a string score entry? I don't see what's the point of it.

image

Found a use case for it, was under my nose this whole time.

**Comment by MaxIsJoe** [comment]
this is ready for review

**Comment by MaxIsJoe** [comment]
hewwo?

File: UnityProject/Assets/Scripts/Objects/Mining/OreRedemptionMachine.cs [hunk 1]
```diff
@@ -114,6 +115,7 @@ public void ClaimLaborPoints(GameObject player)
 				if (itemSlot.ItemObject)
 				{
 					var idCard = AccessRestrictions.GetIDCard(itemSlot.ItemObject);
+					ScoreMachine.AddToScoreInt(laborPoints, "laborPoints");
```
Comment (Bod9001) [inline UnityProject/Assets/Scripts/Objects/Mining/OreRedemptionMachine.cs]: might be worth making a Enum / static reference to the "laborPoints" stringy

Comment (MaxIsJoe) [reply UnityProject/Assets/Scripts/Objects/Mining/OreRedemptionMachine.cs]: will do

**Comment by Bod9001** [comment]
had a quick look looks good

File: UnityProject/Assets/Scripts/Objects/Mining/OreRedemptionMachine.cs (hunk 1 above)
Comment (MaxIsJoe) [reply UnityProject/Assets/Scripts/Objects/Mining/OreRedemptionMachine.cs]: done.

File: UnityProject/Assets/Scripts/HealthV2/Living/LivingHealthMasterBase.cs [hunk 2]
```diff
@@ -1685,6 +1697,14 @@ private void AdminSmash()
 		{
 			AdminCommandsManager.Instance.CmdHealMob(gameObject);
 		}
+
+		private void ClownAbuseScoreEvent(DamageType damageType, GameObject abuser)
+		{
+			if(damageType == DamageType.Clone || damageType == DamageType.Oxy || damageType == DamageType.Radiation) return;
+			if(abuser.TryGetComponent<PlayerScript>(out var script) == false) return;
```
Comment (Alainx277) [inline UnityProject/Assets/Scripts/HealthV2/Living/LivingHealthMasterBase.cs:1702]: Abuser can be `null`, so there should be a check for that.

File: UnityProject/Assets/Scripts/Systems/Score/RoundEndScoreBuilder.cs [hunk 3]
```diff
@@ -0,0 +1,102 @@
...
+		{
+			ScoreMachine.AddNewScoreEntry("laborPoints", "Total Labor Points", ScoreMachine.ScoreType.Int, ScoreCategory.StationScore, ScoreAlignment.Good);
+			ScoreMachine.AddNewScoreEntry("randomEventsTriggered", "Random Events Endured", ScoreMachine.ScoreType.Int, ScoreCategory.StationScore, ScoreAlignment.Good);
+			ScoreMachine.AddNewScoreEntry("foodmade", "Meals Prepared", ScoreMachine.ScoreType.Int, ScoreCategory.StationScore, ScoreAlignment.Good);
+			ScoreMachine.AddNewScoreEntry("hostileNPCdead", "Hostile NPCs dead", ScoreMachine.ScoreType.Int, ScoreCategory.StationScore, ScoreAlignment.Good);
+			ScoreMachine.AddNewScoreEntry("healing", "Healing Done", ScoreMachine.ScoreType.Int, ScoreCategory.StationScore, ScoreAlignment.Good);
+			ScoreMachine.AddNewScoreEntry("foodeaten", "Food Eaten", ScoreMachine.ScoreType.Int, ScoreCategory.StationScore, ScoreAlignment.Weird);
+			ScoreMachine.AddNewScoreEntry("explosions", "Number of explosions this round", ScoreMachine.ScoreType.Int, ScoreCategory.StationScore, ScoreAlignment.Weird);
+			ScoreMachine.AddNewScoreEntry("clownBeaten", "Clown Abused", ScoreMachine.ScoreType.Int, ScoreCategory.StationScore, ScoreAlignment.Bad);
```
Comment (Alainx277) [inline UnityProject/Assets/Scripts/Systems/Score/RoundEndScoreBuilder.cs]: Some of these are defined as constants in other files. Could you define the constants in this class and access them from the other scripts?

File: UnityProject/Assets/Scripts/Systems/Score/ScoreMachine.cs [hunk 4]
```diff
@@ -0,0 +1,183 @@
...
+				Logger.LogError($"[ScoreMachine] - Attempted to add an integer to {ID} but it's entry is not a ScoreEntryInt!");
+				return;
+			}
+			c.Score += valueToAddOnTop;
+		}
+
+		/// <summary>
+		/// whats the new bool for this score?
+		/// </summary>
+		public static void AddToScoreBool(bool newValue, string ID)
```
Comment (Alainx277) [inline UnityProject/Assets/Scripts/Systems/Score/ScoreMachine.cs:97]: Can you use overloading here?

File: UnityProject/Assets/Scripts/Systems/Score/ScoreMachine.cs [hunk 5]
```diff
@@ -0,0 +1,183 @@
...
+			var highestScoreIndex = -1;
+			var winner = "";
+			foreach (var id in Scores.Keys)
+			{
+				if (Scores[id] is not ScoreEntryInt c || c.Score <= highestScoreIndex) continue;
+				highestScoreIndex = c.Score;
+				winner = id;
+			}
+
+			return winner;
```
Comment (Alainx277) [inline UnityProject/Assets/Scripts/Systems/Score/ScoreMachine.cs:169]:
```suggestion
			return ScoreIntWinner(Scores.Keys);
```

File: UnityProject/Assets/Scripts/Systems/Score/ScoreMachine.cs (hunk 4 above)
Comment (MaxIsJoe) [reply UnityProject/Assets/Scripts/Systems/Score/ScoreMachine.cs:97]: Function overloading is the use when you have the same functions that does the same exact thing but with different arguments.
In the future, people might add custom types of scores that have different set of requirements and behaviors, so I've made a conscious decision to not use overloading and make the functions' functionality verbally written, so people can understand what use does these have for every thing.

File: UnityProject/Assets/Scripts/Systems/Score/ScoreMachine.cs (hunk 5 above)
Comment (MaxIsJoe) [reply UnityProject/Assets/Scripts/Systems/Score/ScoreMachine.cs:169]: This function is used for a specific list of stuff that the programmer decides, the other one is used for all scores tracked already.

File: UnityProject/Assets/Scripts/Systems/Score/RoundEndScoreBuilder.cs (hunk 3 above)
Comment (MaxIsJoe) [reply UnityProject/Assets/Scripts/Systems/Score/RoundEndScoreBuilder.cs]: done

File: UnityProject/Assets/Scripts/HealthV2/Living/LivingHealthMasterBase.cs (hunk 2 above)
Comment (MaxIsJoe) [reply UnityProject/Assets/Scripts/HealthV2/Living/LivingHealthMasterBase.cs:1702]: done.

**Comment by MaxIsJoe** [comment]
review pls

File: UnityProject/Assets/Scripts/Systems/Score/ScoreMachine.cs (hunk 5 above)
Comment (Alainx277) [reply UnityProject/Assets/Scripts/Systems/Score/ScoreMachine.cs:169]: Sure, but it does the same thing: get the highest score. Why should that logic be duplicated?

Comment (MaxIsJoe) [reply UnityProject/Assets/Scripts/Systems/Score/ScoreMachine.cs:169]: Error detetction

Comment (Alainx277) [reply UnityProject/Assets/Scripts/Systems/Score/ScoreMachine.cs:169]: You could move the loop without the check into a method and call it from both places, where the public variant first checks all of them against the dictionary.

Whatever man, if you don't want to do it, we'll just leave it.

**Review by Alainx277 — APPROVED** [review]

**Pull request merged**